h2o-core/src/main/java/water/Scope.java

```diff
@@ -20,56 +24,88 @@
     @Override protected Scope initialValue() { return new Scope(); }
   };
-  private final Stack<HashSet<Key>> _keys = new Stack<>();
+  private final Stack<Set<Key>> _keys = new Stack<>();
+  
+  private final Stack<Set<Key>> _protectedKeys = new Stack<>();
   
   /** debugging purpose */
-  static public Scope current() {
+  public static Scope current() {
     return _scope.get();
   }
-
+  
+  /** for testing purpose */
+  public static int currentLevel() {
+    Scope scope = current();
+    assert scope._keys.size() == scope._protectedKeys.size();
+    return scope._keys.size();
+  }
+  
+  /** for testing purpose */
+  public static void reset() {
+    Scope scope = _scope.get();
+    scope._keys.clear();
+    scope._protectedKeys.clear();
+  }
+  
   /** Enter a new Scope */
-  static public void enter() { _scope.get()._keys.push(new HashSet<Key>()); }
+  public static void enter() { 
+    Scope scope = _scope.get();
+    scope._keys.push(new HashSet<>()); 
+    Set<Key> outerProtected = scope._protectedKeys.empty() ? Collections.emptySet() : scope._protectedKeys.peek(); 
+    scope._protectedKeys.push(new HashSet<>(outerProtected)); //inherit protected keys from outer scope
+  }
 
-  /** Exit the inner-most Scope, remove all Keys created since the matching
+  /** Exit the innermost Scope, remove all Keys created since the matching
    *  enter call except for the listed Keys.
    *  @return Returns the list of kept keys. */
-  static public Key[] exit(Key... keep) {
-    List<Key> keylist = new ArrayList<>();
-    if( keep != null )
-      for( Key k : keep ) if (k != null) keylist.add(k);
-    Object[] arrkeep = keylist.toArray();
+  public static Key[] exit(Key... keep) {
+    Scope scope = _scope.get();
+    assert !(scope._keys.empty() || scope._protectedKeys.empty()): "Scope in inconsistent state: Scope.exit() called without a matching Scope.enter()";
+    Set<Key> keepKeys = new HashSet<>();
+    if (keep != null) {
+      for (Key k : keep) {
+        if (k != null) keepKeys.add(k);
+      }
+    }
+    keepKeys.addAll(scope._protectedKeys.pop());
+    Key[] arrkeep = keepKeys.toArray(new Key[0]);
     Arrays.sort(arrkeep);
-    Stack<HashSet<Key>> keys = _scope.get()._keys;
-    if (keys.size() > 0) {
-      Futures fs = new Futures();
-      for (Key key : keys.pop()) {
-        int found = Arrays.binarySearch(arrkeep, key);
-        if (arrkeep.length == 0 || found < 0) Keyed.remove(key, fs, true);
+    Set<Key> removeKeys = scope._keys.pop();
+    Futures fs = new Futures();
+    for (Key key : removeKeys) {
+      boolean remove = arrkeep.length == 0 || Arrays.binarySearch(arrkeep, key) < 0;
+      if (remove) {
+        Value v = DKV.get(key);
+        boolean cascade = !(v == null || v.isFrame()); //Frames are handled differently as we're explicitly also tracking their Vec keys...
+        Keyed.remove(key, fs, cascade);
       }
       fs.blockForPending();
```

Comment:
fs.blockForPending needs to come after the for loop, this is limiting the parallelism of the delete. This is IMHO especially bad together with https://github.com/h2oai/h2o-3/pull/6335#issuecomment-1290668088

Reply:
@michalkurka should be better now.
I added 2 concepts to `Scope`: 
- `Level` to handle more easily and more efficiently the information stored each time we enter a new scope.
- `TrackingInfo` which is just an object storing various info about the keys being tracked. Currently only used with Vecs to have fast access to the `nchunks` from the tracked key. This allows me to bulk remove all vecs with the same `nchunks`.